src/compiler-src.md

```diff
@@ -1,137 +1,198 @@
+`liballoc`, etc. There is also `libproc_macro`, `libtest`, and other runtime
+libraries.
+
+This code is fairly similar to most other Rust crates except that it must be
+built in a special way because it can use unstable features.
+
+## Compiler
+
+> You may find it helpful to read [The Overview Chapter](./overview.md) first,
+> which gives an overview of how the compiler works. The crates mentioned in
+> this section implement the compiler.
+
+The compiler crates all have names starting with `librustc_*`. These are a large
+collection of interdependent crates. There is also the `rustc` crate which is
+the actual binary. It doesn't actually do anything besides calling the compiler
+main function elsewhere.
+
+The dependency structure of these crates is complex, but roughly it is
+something like this:
+
+- `rustc` (the binary) calls [`rustc_driver::main`][main].
+    - [`rustc_driver`] depends on a lot of other crates, but the main one is
+      [`rustc_interface`].
+        - [`rustc_interface`] depends on most of the other compiler crates. It
+          is a fairly generic interface for driving the whole compilation.
+            - The most of the other `rustc_*` crates depend on [`rustc_middle`],
+              which defines a lot of central data structures in the compiler.
+                - [`rustc_middle`] and most of the other crates depend on a
+                  handful of crates representing the early parts of the
+                  compiler (e.g. the parser), fundamental data structures (e.g.
+                  [`Span`]), or error reporting: [`rustc_data_structures`],
+                  [`rustc_span`], [`rustc_errors`], etc.
+
+[main]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/fn.main.html
+[`rustc_driver`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/index.html
+[`rustc_interface`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/index.html
+[`rustc_middle`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/index.html
+[`rustc_data_structures`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/index.html
+[`rustc_span`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/index.html
+[`Span`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html
+[`rustc_errors`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/index.html
+
+You can see the exact dependencies by reading the `Cargo.toml` for the various
+crates, just like a normal Rust crate.
+
+One final thing: [`src/llvm-project`] is a submodule for our fork of LLVM.
+
+Most of this book is about the compiler, so we won't have any further
+explanation of these crates here.
+
+[`src/llvm-project`]: https://github.com/rust-lang/rust/tree/master/src
+
+### Big picture
+
+The dependency structure is influenced strongly by two main factors:
+
+1. Organization. The compiler is a _huge_ codebase; it would be an impossibly
```

Comment:
Out of curiosity I ran tokei on it - 1.7 _million_ lines of code.

Reply:
Did you exclude LLVM? That sounds a bit higher than I remember (~500K), but I might be wrong.

Reply:
That was only counting lines of rust.

<details>

```
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 Alex                    4           85           67            0           18
 Assembly                3         1375          756          402          217
 GNU Style Assembly   6740      1516071       721657       473426       320988
 Autoconf              354        38316        22408        10153         5755
 Automake               12         6065         5149          162          754
 BASH                   10         1791         1387          226          178
 Batch                  18         1598         1342           50          206
 C                    6315       934950       529392       304222       101336
 C Header             7670      1448926       897283       335737       215906
 CMake                1196        59281        47496         5105         6680
 C#                      8          766          571          106           89
 C++                 20444      5295316      3766548       869578       659190
 C++ Header             29        11544         9675          809         1060
 CSS                    43         7378         6041          295         1042
 D                       1           18           16            0            2
 Dockerfile             89         2963         2174          308          481
 .NET Resource           2          269          132          113           24
 Emacs Lisp              9         1296          910          241          145
 FORTRAN Modern          8          398            0          371           27
 GDB Script              1           41           13           15           13
 Go                   1912       482175       362283        71149        48743
 Handlebars              4          138          116            7           15
 HEX                     1           15           15            0            0
 HTML                  186       462384       444556         4440        13388
 INI                     3           30           25            0            5
 JavaScript             55         8000         6409          886          705
 JSON                  144        10028        10028            0            0
 LLVM                23437      3668320      1477481      1865951       324888
 LD Script               1            4            4            0            0
 Lua                     1            7            5            0            2
 Makefile              804         6620         4481          697         1442
 Markdown             1936       171532       171532            0            0
 Module-Definition     118        25860        23250           72         2538
 MSBuild                 8          597          485           88           24
 OCaml                  89        11140         5997         3210         1933
 Objective-C          1721       111733        62368        30723        18642
 Objective-C++         496        37842        24784         6426         6632
 Pascal                 10          316           54          228           34
 Perl                   22         9025         5851         1612         1562
 Protocol Buffers        2          173          118           31           24
 Python               1673       237900       172805        24597        40498
 R                       1           15            7            5            3
 RPM Specfile            1          469          390            0           79
 ReStructuredText     1220       219333       219333            0            0
 Ruby                    1          125           89           23           13
 Rust                19032      1672465      1202002       276415       194048
 Scala                   6           54           54            0            0
 Shell                 246        65387        48964         9122         7301
 SVG                    53         5019         4661          358            0
 Swift                   2           23           17            0            6
 SWIG                   69        10949         7683          512         2754
 TeX                     6         3374         2992           22          360
 Plain Text           1506       517261       517261            0            0
 TOML                  797         8688         7335          179         1174
 TypeScript             25         3784         2906          345          533
 Vim Script             12          593          484           41           68
 Visual Studio Sol|      2           47           47            0            0
 XSL                     3          278          226           14           38
 XML                    51       135907       134752           18         1137
 YAML                  747        73902        59039        12268         2595
-------------------------------------------------------------------------------
 Total               99359     17289959     10993906      4310758      1985295
-------------------------------------------------------------------------------
```

</details>